File: server/app/defense/detectors.py
from __future__ import annotations

import re
from typing import Any

RULES = [
    ("ignore_instructions", re.compile(r"ignore.{0,20}instructions", re.I)),
    ("send_email_tool", re.compile(r"send_email", re.I)),
    ("approve_candidate", re.compile(r"approve_candidate", re.I)),
    ("fake_system_tag", re.compile(r"</?system>", re.I)),
    ("attacker_domain", re.compile(r"@(attacker\.test|evil\.example)", re.I)),
]


def detect_injection_signals(text: str) -> list[dict[str, Any]]:
    hits: list[dict[str, Any]] = []
    for rule, regex in RULES:
        m = regex.search(text)
        if m:
            hits.append(
                {
                    "rule": rule,
                    "index": m.start(),
                    "excerpt": text[max(0, m.start() - 20) : m.start() + 80],
                }
            )
    return hits

File: server/src/defense/detectors.ts
export interface DetectionHit {
  rule: string;
  index: number;
  excerpt: string;
}

const RULES: Array<{ rule: string; re: RegExp }> = [
  { rule: "ignore_instructions", re: /ignore.{0,20}instructions/i },
  { rule: "send_email_tool", re: /send_email/i },
  { rule: "approve_candidate", re: /approve_candidate/i },
  { rule: "fake_system_tag", re: /<\/?system>/i },
  { rule: "attacker_domain", re: /@(attacker\.test|evil\.example)/i },
];

export function detectInjectionSignals(text: string): DetectionHit[] {
  const hits: DetectionHit[] = [];
  for (const { rule, re } of RULES) {
    const m = re.exec(text);
    if (m) {
      hits.push({
        rule,
        index: m.index,
        excerpt: text.slice(Math.max(0, m.index - 20), m.index + 80),
      });
    }
  }
  return hits;
}
